### ISSUE TITLE
Parallize generation of edge-expanded edges

### ISSUE THREAD

**Issue by TheMarex** [Project-OSRM/osrm-backend#3883]
Currently [this section](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/edge_based_graph_factory.cpp#L316-L646) takes about 160 minutes on the OSM Planet on a fairly beefy server. The main problem is that this loop is 100% synchronous and also does IO. By parallelizing this we could cut down processing time by almost ~2.5 hours (160 min -> 10min on our 16 core server).

The main problems that need to be solved for parallelizing are the following:

- `turn_id` is created on-demand here https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/edge_based_graph_factory.cpp#L545
- and implicitly used [here](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/edge_based_graph_factory.cpp#L559-L562) and [here](https://github.com/Project-OSRM/osrm-backend/blob/master/src/extractor/edge_based_graph_factory.cpp#L490-L517) by order
- we can't compute `turn_id` in advance because the number of turns is not known (this is the function that evaluates turn restrictions) and the turn restriction evaluation is conflated in the guidance code.

**Comment by MichalPP** [comment]
could this issue be connected to #3814 ?

**Comment by TheMarex** [comment]
@MichalPP only if your are concerned with the `turn_function` which is the only part of the lua profile that is invoked there.

**Comment by oxidase** [comment]
@TheMarex `turn_id` should not be a big problem, because of a serial bottleneck due to data flushing. It can be done in batches as:
```
for (auto node_at_center_of_intersection = 0; node_at_center_of_intersection < GetNumberOfNodes; node_at_center_of_intersection +=batch_size)
```
1. **parallel** generate tuples `(incoming_edge, turn.eid, OriginalEdgeData)`
2. **serial** flatten TLS vectors
    a. flush `OriginalEdgeData` to the file
    b. assign `turn_id` as tuples `(turn_id, incoming_edge, turn.eid)`
    c. resize `turn_weight_penalties`, `turn_duration_penalties` and `m_edge_based_edge_list`
3. **parallel** generate edge-based edges for every `turn_id, incoming_edge, turn.eid`

The bigger problem is the shared state in  [`TurnLaneHandler`](https://github.com/Project-OSRM/osrm-backend/blob/273fd689ce531ff7eea0b671dc64435ba3b71c6f/src/extractor/edge_based_graph_factory.cpp#L358), [`entry_class_hash`](https://github.com/Project-OSRM/osrm-backend/blob/273fd689ce531ff7eea0b671dc64435ba3b71c6f/src/extractor/edge_based_graph_factory.cpp#L446), [`bearing_class_hash`](https://github.com/Project-OSRM/osrm-backend/blob/273fd689ce531ff7eea0b671dc64435ba3b71c6f/src/extractor/edge_based_graph_factory.cpp#L460) that can make parallelization less efficient.

**Comment by oxidase** [comment]
4 cores CPU utilization  by `osrm-extract` for `germany-latest` ![cpu_germany](https://cloud.githubusercontent.com/assets/4421046/25136825/555e0bf6-2456-11e7-883e-98d2a6ea50ed.png), edge-based edges creation has a time span [1630, 1949] seconds.

**Comment by TheMarex** [comment]
This landed with #4131
